### ISSUE TITLE
Openstack IPI fails to set up image-registry.

### ISSUE THREAD

**Issue by bnordgren** [openshift/installer#4806]
# Version

```console
$ openshift-install version
./openshift-install 4.7.0-0.okd-2021-03-28-152009
built from commit cf35dc04aedf7a80bea5470358025cbe91d232c9
release image quay.io/openshift/okd@sha256:02c01aaaed9d292376e7808b6365ed38c34856e2a13c99718b54a12958f4e6da

```

# Platform:
Openstack
IPI

# What happened?

No web interface, installer emits fatal error message. Failed installation at 99%, with image-registry not initializing. Errors on the image-registry pod indicate that it can't attach or mount a volume. 

## Install-config
```
apiVersion: v1
baseDomain: firelab.org
compute:
- architecture: amd64
  hyperthreading: Enabled
  name: worker
  platform: {}
  replicas: 3
controlPlane:
  architecture: amd64
  hyperthreading: Enabled
  name: master
  platform: {}
  replicas: 3
metadata:
  creationTimestamp: null
  name: contained
networking:
  clusterNetwork:
  - cidr: 10.128.0.0/14
    hostPrefix: 23
  machineNetwork:
  - cidr: 10.1.0.0/16
  networkType: OVNKubernetes
  serviceNetwork:
  - 172.30.0.0/16
platform:
  openstack:
    apiFloatingIP: 192.168.59.13
    apiVIP: 10.1.0.5
    cloud: openstack
    defaultMachinePlatform:
      type: okd.machine
    externalDNS: [ "192.168.59.38", "192.168.100.34" ] 
    externalNetwork: public-net
    ingressVIP: 10.1.0.7
    ingressFloatingIP: 192.168.59.24
publish: External
pullSecret: '{"auths":{"fake":{"auth": "Zm9vOmJhcgo="}}}'
sshKey:
  ssh-ed25519 AAA... bnordgren@mine.usfs-i2.umt.edu
```

## Installer output
```
...
level=debug msg=Still waiting for the cluster to initialize: Working towards 4.7.0-0.okd-2021-03-28-152009: 666 of 669 done (99% complete)
level=debug msg=Still waiting for the cluster to initialize: Working towards 4.7.0-0.okd-2021-03-28-152009: 667 of 669 done (99% complete), waiting on image-registry
level=info msg=Cluster operator baremetal Disabled is True with UnsupportedPlatform: Nothing to do on this Platform
level=info msg=Cluster operator image-registry Available is False with NoReplicasAvailable: Available: The deployment does not have available replicas
level=info msg=ImagePrunerAvailable: Pruner CronJob has been created
level=info msg=Cluster operator image-registry Progressing is True with DeploymentNotCompleted: Progressing: The deployment has not completed
level=error msg=Cluster operator image-registry Degraded is True with Unavailable: Degraded: The deployment does not have available replicas
level=info msg=Cluster operator insights Disabled is True with Disabled: Health reporting is disabled
level=info msg=Cluster operator network ManagementStateDegraded is False with : 
level=error msg=Cluster initialization failed because one or more operators are not functioning properly.
level=error msg=The cluster should be accessible for troubleshooting as detailed in the documentation linked below,
level=error msg=https://docs.openshift.com/container-platform/latest/support/troubleshooting/troubleshooting-installations.html
level=error msg=The 'wait-for install-complete' subcommand can then be used to continue the installation
level=fatal msg=failed to initialize the cluster: Working towards 4.7.0-0.okd-2021-03-28-152009: 667 of 669 done (99% complete), waiting on image-registry
```
## oc get clusteroperators
```
NAME                                       VERSION                         AVAILABLE   PROGRESSING   DEGRADED   SINCE
authentication                             4.7.0-0.okd-2021-03-28-152009   True        False         False      77m
baremetal                                  4.7.0-0.okd-2021-03-28-152009   True        False         False      106m
cloud-credential                           4.7.0-0.okd-2021-03-28-152009   True        False         False      115m
cluster-autoscaler                         4.7.0-0.okd-2021-03-28-152009   True        False         False      105m
config-operator                            4.7.0-0.okd-2021-03-28-152009   True        False         False      106m
console                                    4.7.0-0.okd-2021-03-28-152009   True        False         False      84m
csi-snapshot-controller                    4.7.0-0.okd-2021-03-28-152009   True        False         False      105m
dns                                        4.7.0-0.okd-2021-03-28-152009   True        False         False      103m
etcd                                       4.7.0-0.okd-2021-03-28-152009   True        False         False      104m
image-registry                                                             False       True          True       98m
ingress                                    4.7.0-0.okd-2021-03-28-152009   True        False         False      89m
insights                                   4.7.0-0.okd-2021-03-28-152009   True        False         False      99m
kube-apiserver                             4.7.0-0.okd-2021-03-28-152009   True        False         False      102m
kube-controller-manager                    4.7.0-0.okd-2021-03-28-152009   True        False         False      103m
kube-scheduler                             4.7.0-0.okd-2021-03-28-152009   True        False         False      102m
kube-storage-version-migrator              4.7.0-0.okd-2021-03-28-152009   True        False         False      88m
machine-api                                4.7.0-0.okd-2021-03-28-152009   True        False         False      99m
machine-approver                           4.7.0-0.okd-2021-03-28-152009   True        False         False      105m
machine-config                             4.7.0-0.okd-2021-03-28-152009   True        False         False      104m
marketplace                                4.7.0-0.okd-2021-03-28-152009   True        False         False      104m
monitoring                                 4.7.0-0.okd-2021-03-28-152009   True        False         False      87m
network                                    4.7.0-0.okd-2021-03-28-152009   True        False         False      106m
node-tuning                                4.7.0-0.okd-2021-03-28-152009   True        False         False      104m
openshift-apiserver                        4.7.0-0.okd-2021-03-28-152009   True        False         False      99m
openshift-controller-manager               4.7.0-0.okd-2021-03-28-152009   True        False         False      103m
openshift-samples                          4.7.0-0.okd-2021-03-28-152009   True        False         False      96m
operator-lifecycle-manager                 4.7.0-0.okd-2021-03-28-152009   True        False         False      104m
operator-lifecycle-manager-catalog         4.7.0-0.okd-2021-03-28-152009   True        False         False      104m
operator-lifecycle-manager-packageserver   4.7.0-0.okd-2021-03-28-152009   True        False         False      99m
service-ca                                 4.7.0-0.okd-2021-03-28-152009   True        False         False      106m
storage                                    4.7.0-0.okd-2021-03-28-152009   True        False         False      103m
```
## oc describe pod {image-registry}

```
Events:
  Type     Reason              Age                  From                     Message
  ----     ------              ----                 ----                     -------
  Warning  FailedScheduling    102m                 default-scheduler        0/3 nodes are available: 3 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate.
  Warning  FailedScheduling    102m                 default-scheduler        0/3 nodes are available: 3 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate.
  Warning  FailedScheduling    102m                 default-scheduler        0/3 nodes are available: 3 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate.
  Warning  FailedScheduling    99m                  default-scheduler        0/3 nodes are available: 3 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate.
  Warning  FailedScheduling    96m                  default-scheduler        0/5 nodes are available: 2 node(s) had taint {node.kubernetes.io/not-ready: }, that the pod didn't tolerate, 3 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate.
  Warning  FailedScheduling    95m                  default-scheduler        0/6 nodes are available: 3 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate, 3 node(s) had taint {node.kubernetes.io/not-ready: }, that the pod didn't tolerate.
  Normal   Scheduled           94m                  default-scheduler        Successfully assigned openshift-image-registry/image-registry-668544d68f-frv9k to contained-nrw2c-worker-0-lc49b
  Warning  FailedMount         40m (x5 over 92m)    kubelet                  Unable to attach or mount volumes: unmounted volumes=[registry-storage], unattached volumes=[bound-sa-token registry-token-kvggf registry-storage registry-tls ca-trust-extracted registry-certificates trusted-ca installation-pull-secrets]: timed out waiting for the condition
  Warning  FailedMount         36m (x4 over 75m)    kubelet                  Unable to attach or mount volumes: unmounted volumes=[registry-storage], unattached volumes=[installation-pull-secrets bound-sa-token registry-token-kvggf registry-storage registry-tls ca-trust-extracted registry-certificates trusted-ca]: timed out waiting for the condition
  Warning  FailedMount         22m (x4 over 71m)    kubelet                  Unable to attach or mount volumes: unmounted volumes=[registry-storage], unattached volumes=[trusted-ca installation-pull-secrets bound-sa-token registry-token-kvggf registry-storage registry-tls ca-trust-extracted registry-certificates]: timed out waiting for the condition
  Warning  FailedMount         14m (x7 over 84m)    kubelet                  Unable to attach or mount volumes: unmounted volumes=[registry-storage], unattached volumes=[ca-trust-extracted registry-certificates trusted-ca installation-pull-secrets bound-sa-token registry-token-kvggf registry-storage registry-tls]: timed out waiting for the condition
  Warning  FailedMount         11m (x6 over 67m)    kubelet                  Unable to attach or mount volumes: unmounted volumes=[registry-storage], unattached volumes=[registry-tls ca-trust-extracted registry-certificates trusted-ca installation-pull-secrets bound-sa-token registry-token-kvggf registry-storage]: timed out waiting for the condition
  Warning  FailedMount         9m55s (x3 over 46m)  kubelet                  Unable to attach or mount volumes: unmounted volumes=[registry-storage], unattached volumes=[registry-certificates trusted-ca installation-pull-secrets bound-sa-token registry-token-kvggf registry-storage registry-tls ca-trust-extracted]: timed out waiting for the condition
  Warning  FailedMount         5m48s (x5 over 90m)  kubelet                  Unable to attach or mount volumes: unmounted volumes=[registry-storage], unattached volumes=[registry-token-kvggf registry-storage registry-tls ca-trust-extracted registry-certificates trusted-ca installation-pull-secrets bound-sa-token]: timed out waiting for the condition
  Warning  FailedAttachVolume  4m8s (x52 over 94m)  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-5b346b2c-ea7f-4a27-9bb3-6bfa84c49284" : status of volume 5170d1f8-6403-42b2-a942-5e31e53d5ea0 is error
  Warning  FailedMount         101s (x9 over 88m)   kubelet                  Unable to attach or mount volumes: unmounted volumes=[registry-storage], unattached volumes=[registry-storage registry-tls ca-trust-extracted registry-certificates trusted-ca installation-pull-secrets bound-sa-token registry-token-kvggf]: timed out waiting for the condition
```

# What you expected to happen?

Complete install. 

# How to reproduce it (as minimally and precisely as possible)?

```console
$ ./openshift-install create cluster --dir=<blah> --log-level=debug
```

# Anything else we need to know?

It became hung up at this same spot three times. Occasionally it fails earlier in the process, but it never gets past this. I did let it sit overnight once, in the hopes that it would finish initializing. No joy. 

Openstack cluster is on prem behind a NAT. DNS entries for cluster on firelab.org is accomplished by 192.168.59.38 above, and is functional (would assume it would fail much earlier if DNS weren't working.)

**Comment by staebler** [comment]
Now you are asking questions that I don't have any answers for. My OpenStack knowledge is limited.

**Comment by bnordgren** [comment]
So, if I read [this page](https://docs.okd.io/latest/storage/dynamic-provisioning.html#openstack-cinder-storage-class_dynamic-provisioning) correctly, I need to do this: 

```
[root@079a0592c5e1 ~]# ./oc patch storageclass standard -p '{"parameters": { "availability" : "storage0"}}'
The StorageClass "standard" is invalid: parameters: Forbidden: updates to parameters are forbidden.
```

But apparently I cannot. There's also a "standard" and a "standard-csi", but I don't know the difference and I haven't found documentation on the parameters taken by "standard-csi". How would one add the above parameter to the following description in such a way that the installation completes? 

```
[root@079a0592c5e1 ~]# ./oc describe storageclass standard
Name:                  standard
IsDefaultClass:        Yes
Annotations:           storageclass.kubernetes.io/is-default-class=true
Provisioner:           kubernetes.io/cinder
Parameters:            <none>
AllowVolumeExpansion:  True
MountOptions:          <none>
ReclaimPolicy:         Delete
VolumeBindingMode:     WaitForFirstConsumer
Events:                <none>
```

**Comment by bnordgren** [comment]
Ah. This is a "[known issue](https://github.com/openshift/installer/blob/5379519a0beb92d39227239cca52788caa7d0caa/docs/user/openstack/known-issues.md)". Scroll down to "Cinder Availability Zone". 

Having followed the advice there, and furthermore making my new sc-availability storageclass the default, I started a `openstack-install wait-for install-complete`. No forward progress, still at "667 of 669 done". 

I noticed the old volume still hanging out (where "`oc get pv`" showed a volume with no obvious error,  and the Horizon web interface showing the volume with an error.) I deleted the volume using the Horizon web interface, and this is where I think I went wrong. `oc get pv` still showed the volume. `oc delete pv <id>` now just hangs. The volume is there with state "Terminating".  I suspect I probably should have deleted the volume using `oc` or `kubectl`. 

So what I'm learning from this is that kubernetes/okd does not query for the state of volumes from the provider, it just assumes that: a] all creation requests succeed; and b] volumes are not manipulated by any other means. Item 'b' is probably a reasonable constraint, but item 'a' likely needs fixing.

The question remains, tho. How do I get the install to succeed? Is there a way to provide the workaround to install-config.yaml? If not, how do I restart the installation after I manually perform the steps above?

**Comment by bnordgren** [comment]
Attempting to force delete the pv with no backing volume in cinder emits a warning that it removes the resource from the API immediately and does not wait for confirmation...and yet hangs anyway.

```
[root@079a0592c5e1 ~]# ./oc delete --force pv pvc-5b346b2c-ea7f-4a27-9bb3-6bfa84c49284
warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
persistentvolume "pvc-5b346b2c-ea7f-4a27-9bb3-6bfa84c49284" force deleted
```

**Comment by bnordgren** [comment]
Tried again with a clean slate (destroyed cluster, created cluster, let it hang, performed [workaround on known issue page](https://github.com/openshift/installer/blob/5379519a0beb92d39227239cca52788caa7d0caa/docs/user/openstack/known-issues.md)). This time, I tried deleting the pv using the `oc` command instead of Openstack Cinder. There was a pause for a very long time (>5 minutes) and then an error message ("client connection lost"):

```
[root@079a0592c5e1 ~]# ./oc delete pv pvc-6299ff0f-b37f-4f92-8c84-ae19fcb5ef80
persistentvolume "pvc-6299ff0f-b37f-4f92-8c84-ae19fcb5ef80" deleted
error: An error occurred while waiting for the object to be deleted: an error on the server ("unable to decode an event from the watch stream: http2: client connection lost") has prevented the request from succeeding
```

After another very long time (>5 minutes): 
```
Error from server: etcdserver: leader changed
```
and the command terminates. The volume is still there in cinder and in the `./oc get pv` list (with a status of Terminating). 

Repeating the `oc` command, but including the `--force` flag shows exactly the same behavior and results.

Big picture thus far: The identified workaround for Cinder AZs cannot be used after the IPI installer paints itself into a corner. Once the installer grinds to a halt, there is no visible means of getting it moving again (although I am willing to be shown to be wrong on this last point.) As far as I can tell, the only paths forward require that the IPI avoid creating a volume in an error state for the image registry, or a means of manually recovering from an error state be created. Ideally, a parameter in `install-config.yaml` would allow the user to specify a specific AZ, and if that is present, the installer would perform the steps in the Known Issues document instead of creating a storageclass object lacking this information. Alternatively, creating a way to recover from a PV with an error state may allow manual intervention to succeed. I do not know which path is less effort, although both items seem worth pursuing.
